Retrieve fish length parameters

langbart opened the pull request:
Hi @efcaguab , I have created a couple of basic functions to extract fish length parameters. I preferred to use airtable's fao_catch table to extract taxonomic information and integrate it with FishBase info through the `rfishbase` package. The extraction of parameters for single species is quite immediate (see `length_params_sp()`), whereas for groups composed of several species (see `length_params_spp`) perhaps a criterion should be defined. I'm not sure that averaging all the length parameters between different species (even if of the same genus) is correct, at the moment I'm taking the length parameters from the most represented species, what do you think?

efcaguab commented:
Awesome. I agree that averaging parameters across species is not ideal. Initially I thought about limiting it to species within the taxon that might be present in Timor as a way to mitigate that, but yeah not ideal. 

Using only the species that are fished would be great! I'm not sure where we could get that information though. 

The only thing that we could potentially use is the first survey. Do you think that's usable? And if yes do you think the quality would be good enough? I'm not sure the species identification was very good back then (snappers and emperors can be challenging for the untrained eye) and in any case many of the older "species" categories were at the genus or species level as well.

langbart commented:
Hi @efcaguab  I changed some stuff. The final output of these functions is a table with the length parameters of each FAO code. It remains to understand how to derive the lengths from groups composed of many species. Maybe we could do some statistics and evaluate. For example, we could calculate the deviation of the length parameters for each group and evaluate whether it is reasonable to use the measurements of a single "type" species for each group. Another point to ponder on is that there are 12 groups missing from the FishBase database associated with Timor. I think the best thing is to focus on them manually. I don't know if I've made it clear,  we can eventually have a catch up on this.

efcaguab commented on Dockerfile.prod:
```diff
@@ -36,6 +36,8 @@ RUN install2.r --error --skipinstalled \
     rlang \
     furrr \
     future
```
```suggestion
    future \
```

efcaguab commented on R/length-params.R:
```diff
@@ -0,0 +1,158 @@
+    dplyr::bind_cols(spp,.) %>%
+    dplyr::mutate(name=name_scientific) %>%
+    dplyr::mutate(name=dplyr::case_when(name=="Thunnini"~"Tuna",
+                                        name=="Rajiformes"~"Manta",
```
I don't think Mantas are Rajiformes. Actually this makes me think that we might have chosen the wrong taxocode here as Rajiformes are skates but I would believe that stingrays are also often captured in Timor. Do you have any insight @altilley?

langbart replied:
Yes @efcaguab that's a point I didn't highlight, happy that you took it. There is some confusion on the fao_catch table, do unnamed record rows in the fao_catch table should be removed?

efcaguab commented on R/length-params.R:
```diff
@@ -0,0 +1,158 @@
+
+  catch_types <- metadata$catch_types
+
+  fao_records <- air_get_records(table="fao_catch",
```
I think it would be best if we add this to the ingest metadata tables step so we get both tables at once from the metadata and we can ensure that we're using the same version of the metadata throughout a run of the pipeline

efcaguab commented on R/length-params.R:
```diff
@@ -0,0 +1,158 @@
+
+  catch_types <- metadata$catch_types
+
+  fao_records <- air_get_records(table="fao_catch",
+                                 base_id=pars$metadata$airtable$base_id,
+                                 api_key=readLines("auth/airtable-key")) %>%
+    air_records_to_tibble()
```
I think it would be best if we add this to the ingest metadata tables step so we get both tables at once from the metadata and we can ensure that we're using the same version of the metadata throughout a run of the pipeline.

efcaguab commented on R/length-params.R:
```diff
@@ -0,0 +1,158 @@
+      dplyr::slice(1)
+  }
+
+  if (rank=="Species"){
+    specs_length <- rfishbase::length_length(taxa)
+    return(specs_length)
+  }
+
+  if (rank=="Genus"){
+    taxa <- stringr::word(taxa, 1)
+    specs <-
+      rfishbase::country(rfishbase::species_list(Genus=taxa)) %>%
+      dplyr::filter(C_Code==country_code) %>%
+      magrittr::extract2("Species")
+  }
+
+  if (rank=="Family"){
+    specs <-
+      rfishbase::country(rfishbase::species_list(Family=taxa)) %>%
+      dplyr::filter(C_Code==country_code) %>%
+      magrittr::extract2("Species")
+  }
+
+  if (rank=="comm_name"){
+    sp_list <-
+      rfishbase::common_to_sci(taxa) %>% magrittr::extract2("Species")
+    specs <-
+      rfishbase::country(sp_list) %>%dplyr::filter(C_Code==country_code)%>%
+      magrittr::extract2("Species")
+  }
```
I haven't tested it, but I think you could avoid most of the if statements (except the common name one) by executing a function call

```suggestion
  if (rank == "comm_name") {
    sp_list <- rfishbase::common_to_sci(taxa) %>%
      magrittr::extract2("Species")
  } else {
    sp_list <- list(taxa) %>%
      rlang::set_names(rank) %>%
      do.call(rfishbase::species_list, .)
  }

  specs <- sp_list %>%
    rfishbase::country() %>%
    dplyr::filter(C_Code %in% country_code) %>%
    magrittr::extract2("Species")
```

langbart replied:
Yeah @efcaguab, that's a great tip!

efcaguab commented on R/length-params.R:
```diff
@@ -0,0 +1,158 @@
+    sp_len <-
+      get_fish_length(taxa=dat$name,
+                      rank=dat$taxa_rank,
+                      country_code=country_code,
```
Because there are not many museum specimens from Timor, I think we should also include Indonesia and Australian specimens. That might help solve some of the gaps we have for some taxa.

langbart replied:
Yeah, I have included Indonesia and Timor as default, unfortunately, the code for Australia pick all the Australia, that's probably too much. Any way we can easily add it and see if it helps

efcaguab commented:
Good stuff @langbart. Thanks! I've made a couple of suggestions to the code. Thanks also for documenting the functions so well. It was really easy to dive into your code. 

> It remains to understand how to derive the lengths from groups composed of many species. Maybe we could do some statistics and evaluate. For example, we could calculate the deviation of the length parameters for each group and evaluate whether it is reasonable to use the measurements of a single "type" species for each group. 

This is a good idea. And it's perhaps something we can chat about in the meeting tomorrow. I'm sure @altilley might have some ideas about the best way to go around this issue. 

> Another point to ponder on is that there are 12 groups missing from the FishBase database associated with Timor. I think the best thing is to focus on them manually. 

I think it's OK if we include species observed in Indo and Australia (especially Northern Australia if that's at all possible) because there are not that many specimens from Timor (which is where I think FishBase gets this info from)

efcaguab commented:
Also I'm thinking that maybe we should consider adopting a common style guide. I was thinking the [tidyverse guide](https://style.tidyverse.org) could be a good starting point. Maybe not for this, but for future pull requests. What do you think?

langbart commented:
Thanks for the feedback @efcaguab ! I changed a bit of things according to your great tips. There is still something to fix (I get some warnings using `retrieve_lengths`) which I think depends on some stuff to fix in the catch tables (`fao_catch` and `catch_types`) in airtable, we can discuss that on the catch-up. For what concerns the common style guide I definitely agree!

langbart commented on R/validate-landings.R:
```diff
@@ -116,7 +132,21 @@ validate_landings <- function(log_threshold = logger::DEBUG){
+    tidyr::unnest(.data$length_frequency, keep_empty = TRUE) %>%
+    dplyr::left_join(weight_table) %>%
+    dplyr::mutate(tot_weight = (.data$a * .data$length^.data$b) * individuals) %>%
+    tidyr::nest(length_frequency = c(.data$length:.data$tot_weight)) %>%
+    dplyr::mutate(tot_weight = purrr::map_dbl(.data$length_frequency, ~ {
+      sum(.x$tot_weight, na.rm = TRUE)
+    })) %>%
+    tidyr::nest(landed_catch = c(.data$catch_taxon:.data$tot_weight)) %>%
+    dplyr::mutate(tot_weight = purrr::map_dbl(.data$landed_catch, ~ {
+      sum(.x$tot_weight, na.rm = TRUE)
```
@efcaguab I really don't like that, I'm sure you have an easy purrr-way to do it smarter

langbart replied:
@efcaguab I really don't like that way to summarise the total weight from the nested structure (and I think it also causes troubles). I'm sure you have an easy purrr way to get it smarter

langbart commented:
Hi @efcaguab  I checked the dispersion of the coefficients a and b associated with each group of species. I changed the search name of some species groups in order to have estimates of a and b closer to the actual catches. Fortunately, the `rfishbase` package provides some coefficient accuracy measurements that greatly reduce uncertainty. There are a couple of issues to resolve:

-  I could not get the coefficients associated with the "FLY" group, they are not on FishBase. Similarly, there are other groups such as crustaceans and molluscs which have no coefficients. How we can deal with them? Do you think adding manually from other sources is the right way?

- I added a column summarizing the total weight of each trip in `validate_landings`. Do you think it fits there or is it better to put it in the preprocessing pipeline?

efcaguab commented on R/length-params.R:
```diff
@@ -1,68 +1,165 @@
+      name = dplyr::case_when(
+        name == "Thunnini" ~ "Tuna",
+        name == "Selachimorpha" ~ "Shark",
+        name == "Hyporhamphus quoyi" ~ "Garfish",
+        name == "Abudefduf vaigiensis" ~ "Sergeant",
```
Wondering why we need to do this? Is it because Fishbase is struggling to find a/b parameters for these species or because we we want to get parameters for other Garfishes and Sergeant fishes?

efcaguab commented on R/length-params.R:
```diff
@@ -1,68 +1,165 @@
+        name == "Selachimorpha" ~ "Shark",
+        name == "Hyporhamphus quoyi" ~ "Garfish",
+        name == "Abudefduf vaigiensis" ~ "Sergeant",
+        catch_name_en == "Unicornfish" ~ "Unicornfish",
+        catch_name_en == "Surgeonfish" ~ "Surgeonfish",
+        catch_name_en == "Butterflyfish" ~ "Butterflyfish",
+        catch_name_en == "Bannerfish" ~ "Bannerfish",
```
I might be missing something but I think we don't need this. The unit of analysis is the FAO groups and so I think is best if we stick to that as much as possible. I think it'll be OK to have Unicorn and surgeonfishes as well as Butterfly and Bannerfish togetert when we get the a/b parameters.

langbart commented:
Hey @efcaguab, I updated the code according to your review, they were basically experiments and they were not supposed to be there. Anyway, the thing that should still be discussed is point 1 in my previous comment. Moreover, do we need to create a sheet in metadata airtable with length-weight info?

efcaguab commented on R/length-params.R:
```diff
@@ -0,0 +1,155 @@
+    dplyr::filter(!.data$catch_name_en %in% c(
+      "No catch", "Sea cucumber",
+      "Other", "Unknown", "Lobster",
+      "Herring"
```
Perhaps we should also exclude octopus, shrimps, cuttlefish, crabs, cockles, and algae? Also wondering why are herrings excluded here?

efcaguab commented:
> 1. I could not get the coefficients associated with the "FLY" group, they are not on FishBase. Similarly, there are other groups such as crustaceans and molluscs which have no coefficients. How we can deal with them? Do you think adding manually from other sources is the right way?

- Not sure what to do about Flying fish. One alternative is to just comb the literature manually and include some a/b parameters for this group.
- For Lobster, Seaweed, and Sea cucumber I think we don't need to worry as it looks like data about these species is not collected (they don't have a code in the survey) but we need to check. 
- Octopus, shrimps, cuttlefish, crabs, cockles, and algae also need special treatment and in the past I think we just used an average weight per individual. 
- For flying fish we might need to comb the literature and manually get a/b parameters. 
- I noticed stingrays are also missing from the data. We might need to look into that as well. We still need to figure out wether the Airtable should be updated so that We have BAI (Batoidea) or just SRX. 

I say for now we can just exclude these species from the calculations. We can create new issues that we can fix at a later date. We can talk about it in the meeting later today. Perhaps is something Alex can give us a hand with so we can focus on the coding. 

> 2. I added a column summarizing the total weight of each trip in `validate_landings` (see also [Retrieve fish length parameters #48 (comment)](https://github.com/WorldFishCenter/peskas.timor.data.pipeline/pull/48#discussion_r663529042)). Do you think it fits there or is it better to put it in the preprocessing pipeline?

I think that adding the weight should definitively happen outside and before the validation. In part because it's not really validation code and also because ideally we want to validate these weights and also use them to validate other fields. I think that a good idea would be to add two workflow jobs here. 
1. One job to get the parameters from fish base and get a "clean" data frame with a-b parameters of each study. Each row is an individual a/b parameter. The output can go to the google drive. I don't think we need taxon level a-b data at this stage in airtable. 
2. Another job that retrieves the output of 1 from the storage and adds the weights to the merged landings. The validation step can then use the output of this step as the input. 

Regarding the weight calculation I think we need a bit more work before we're done with that.  I plotted the l/w relationships for the species we have info on. Colours correspond to certain type of length and the black line is the relationship generated using the median of a, and b. 

<img width="964" alt="Screen Shot 2021-07-07 at 5 32 33 PM" src="https://user-images.githubusercontent.com/8033969/124705345-a47a0980-df49-11eb-9c50-b214a0f440fa.png">

Some observations: 

- There is quite a bit of difference on wether we use total, fork, standard, or other length. We need to check what's the one used in the field and convert the relationships to the ones used (when possible) by using the length_length table in rfishbase. If not possible I think we can still use them. Hopefully the right ones will outnumber the wrong ones. Note that for some studies this info is already available as aTL and no extra conversion is needed. 
- Using the median seems to get rid of the very weird relationships so I think it's fine to include those with a coefficient of determination below 0.95.

langbart commented:
Hey @efcaguab , I've updated some code according to your and Alex's suggestions. Now the table with the length-weight parameters (rfish_table) is on the cloud and is downloaded to add the weight column in the `merge_landings` function and no longer in `validate_landings`. I did an extensive search on SeaLifeBase and added the morphometric parameters of the groups that were missing (https://docs.google.com/spreadsheets/d/1vaS0jYz4NGJbHFXV8A6YjLegmRpIOOAbi3TL38_seps/edit#gid=0) and selected for each a standard length measurement (for example the mantel length (ML) for cephalopods, the shell length (Shl) for bivalves and so on), although, as for fishes, it takes a double-check from Alex.

langbart commented:
Hey @efcaguab , I've updated some code according to your and Alex's suggestions. Now the table with the length-weight parameters (rfish_table) is on the cloud and is downloaded to add the weight column in the `merge_landings` function and no longer in `validate_landings`. I did an extensive search on SeaLifeBase and added the morphometric parameters of the groups that were missing (https://docs.google.com/spreadsheets/d/1vaS0jYz4NGJbHFXV8A6YjLegmRpIOOAbi3TL38_seps/edit#gid=0) and selected for each a standard length measurement (for example the mantel length (ML) for cephalopods, the shell length (Shl) for bivalves and so on), although, as for fishes, it takes a double-check from Alex.

efcaguab commented on R/length-params.R:
```diff
@@ -74,51 +68,44 @@ get_catch_types <- function(log_threshold = logger::DEBUG) {
+      do.call(rfishbase::species_list, .)
+  }
+
+  if (taxa == "Abudefduf vaigiensis" | taxa == "Hyporhamphus quoyi") {
+    specs_length <- rfishbase::length_weight(taxa)
+    return(specs_length)
```
Just curious about why there is particular treatment for these two species?

langbart replied:
The problem with these species was that they hadn't a country code assigned in the table returned by `rfishbase::country()` although they occur all around Timor and Indonesia.
https://www.fishbase.se/summary/Hyporhamphus-quoyi.html
https://www.fishbase.se/summary/Abudefduf-vaigiensis.html

efcaguab replied:
Thanks for checking this too! It might be worth to add a comment in the code to explain why is this the case.

efcaguab commented on R/length-params.R:
```diff
@@ -42,30 +38,28 @@ get_catch_types <- function(log_threshold = logger::DEBUG) {
         name == "Thunnini" ~ "Tuna",
         name == "Selachimorpha" ~ "Shark",
+        name == "Reptantia" ~ "Nephropidae",
+        name == "Rajiformes" ~ "Myliobatiformes",
```
So I had another look and looks like ISSCAAP uses an old taxonomy in which the order Rajiformes encompasses the current Rajiformes as well as Myliobatiformes. Cause they might also catch skates we cannot just exclude Rajiformes I think. Maybe we need to duplicate a row in the spp data frame by hand so that we get the species for both of these orders? Something like:

```
spp <- ... %>%
    dplyr::add_row(
      interagency_code = "SRX",
      isscaap_taxocode.x = "110XXXXXXX",
      name_scientific = "Myliobatiformes"
    )
```

If we do that we also need to remove the `dplyr::filter(dplyr::row_number() == 1)` at the end of this function and modify `retrieve_lengths()` so that it iterates by name and not by interagency code

langbart replied:
Here I replaced with Myliobatiformes because Rajiformes didn't occur in Timor and Indonesia according to the FishBase database. It sounded strange to me but effectively Rajiformes (and Rajidae, the skates) are rare in tropical shallow waters: https://www.fishbase.se/summary/OrdersSummary.php?order=Rajiformes. Most of Batoidea in Timor are Myliobatiformes

efcaguab replied:
Oh awesome!!! That's perfect! Thanks for checking!

efcaguab commented on R/length-params.R:
```diff
@@ -42,30 +38,28 @@ get_catch_types <- function(log_threshold = logger::DEBUG) {
       name = dplyr::case_when(
         name == "Thunnini" ~ "Tuna",
         name == "Selachimorpha" ~ "Shark",
+        name == "Reptantia" ~ "Nephropidae",
```
Pending check with Alex to see whether most of the lobsters caught in East Timor are spiny lobsters or actual lobsters

langbart replied:
Just updated the LOX group with _Panulirus_ species

efcaguab replied:
Perfect! Just as with the other manual hacks it might be good to add a comment to the code. I find comments specially useful to explain *why* a piece of code unexpected code has been written.

langbart replied:
Yeah you are completely right, sorry for that!

efcaguab commented:
Awesome Lorenzo! Great progress. Amazing to see the manual collection for the species in SeaLife base, must have been a lot of work!

Couple of comments: 

- Would be good to move the information you collected from the google sheets to the metadata base in Airtable. That way we have all the metadata in one place. 
- The only major thing I think we're missing is the TL/FL for fishes. Right now we use the median of a parameters across TL, FL, and other lengths. We cannot really do that. We need to get the median for each of them. It would be possibly good to have an a_tl and an a_fl columns in rfish_table which then we can use for the weight calculation accordingly (but this will only be aplicable for fish :( ). I think we can  convert a_tl to a_fl and visceversa using the length_length tables.

langbart commented:
Thanks @efcaguab , the manual collection was a bit time consuming but it was very interesting to take a look at the Timor fauna. In any case, for the first point, I proceeded to put the morphometric data in airtable (https://airtable.com/tblhM0X1MHv9G2eYz/viw1ouq8TFWwfRbDU?blocks=hide). As for the second point, the length on which the median is calculated is TL for all fish groups except for Myliobatiformes, which uses the WD (width) (you can see it in the `join_weights` function)

langbart commented:
Hi Fer, you were right, median was calculated on all lengths types! I fixed the code, now `merged_landings` includes the medians of the different lengths types separately. However, while we have the length-weight coefficients associated with TL for all groups (fishes), some groups lack the length-weight coefficients associated with FL: DRZ, FLY, RAB, DDD, SFA, MOO, MHL, MSD, KAH, GZP. Each of these groups is composed of one species, so a literature check should be the best thing. What do you think?

efcaguab commented:
Mate that's awesome! Really smart about going up to the genus level for these 4 species you mentioned. I think it does make total sense both from the data-collection and the data-processing perspective. 

After that change, how many taxa are there now left that might still needing manual literature review? 

Let me know if you think I should have a look at the latest changes you made or if it's better I wait until the weight calculation it's been done.

langbart commented:
Hey @efcaguab, the situation is:
All groups have weight info in TL except "SRX" (stingrays and similar which are generally measured in WD) and non-fish groups.
All groups have also weight info in FL except the non-fish groups and MHL, FLY, DRZ. However, there is no need to convert DRZ as it has TL = FL (that's because it's a very strange but cute fish https://www.fishbase.se/photos/PicturesSummary.php?StartRow=3&ID=454&what=species&TotRec=7 ). As for MHL and FLY I will do a new literature search although I believe that, at least in MHL, the use of TL or FL is quite irrelevant (see its morphology https://www.fishbase.se/photos/ThumbnailsSummary. php? ID = 5802).

So I will try to get more from the literature about SRX, MHL and FLY and then calculate the final weight.

langbart commented:
Hey @efcaguab, I have added the weight values. At the moment the weight column is in the nested list `length_individuals`, I was trying to extract the cumulative sum of weights from the nested lists but without success... Do you have a tip (probably some purrr-like function) to solve that?

efcaguab commented:
One easy option could be to calculate the total weight before renesting the dataset? If it needs to happen after it's been nested one could use a function and purrr. Something like the code below could do the trick.

```r
calc_total_weight <- function(x){
    x$length_individuals %>%
      purrr::map_dbl(~ sum(.$weight, na.rm = T)) %>%
      sum()
  }

data %>% 
    dplyr::mutate(catch_weight = purrr::map(.data$species_group, calc_total_weight))
```

That said, looks like the nesting might be more cumbersome than helpful. So this is something we could think about removing in the future

efcaguab commented:
Awesome! I'll have a careful look tomorrow!

efcaguab commented:
@langbart I made a few changes but didn't manage to check the whole thing. I tried to add a description to each of the commits I did so you know what I did and why. Let me know if you have any question so far or if you spot anything wrong/weird in what I made. Hope to finish early next week! Cheers!

langbart commented:
@efcaguab thanks for your review, I was just wondering how to deal with the morphometric table in airtable. At the moment I'm "manually" filling this table as it composed of the rifsh_table generated from `retrieve_lengths` and another table that I'm manually filling from literature. Do you think that it would be better to add another step into pipeline to merge these two tables and ingest them into the airtable directly?

efcaguab commented:
Yes. I think it’ll be best to merge. The reason is mainly reproducibility. First, if we add more species the code wouldn’t really work unless we update the one in Airtable. Second, merging it also gives us flexibility to change the approach if needed.  Lastly it could be hard to know when or wether the Airtable has the updated info from retrieve lengths. 

> On 31/07/2021, at 03:50, langbart ***@***.***> wrote:
> 
> ﻿
> @efcaguab thanks for your review, I was just wondering how to deal with the morphometric table in airtable. At the moment I'm "manually" filling this table as it composed of the rifsh_table generated from retrieve_lengths and another table that I'm manually filling from literature. Do you think that it would be better to add another step into pipeline to merge these two tables and ingest them into the airtable directly?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

langbart commented:
> Yes. I think it’ll be best to merge. The reason is mainly reproducibility. First, if we add more species the code wouldn’t really work unless we update the one in Airtable. Second, merging it also gives us flexibility to change the approach if needed. Lastly it could be hard to know when or wether the Airtable has the updated info from retrieve lengths.
> […](#)
> On 31/07/2021, at 03:50, langbart ***@***.***> wrote: ﻿ @efcaguab thanks for your review, I was just wondering how to deal with the morphometric table in airtable. At the moment I'm "manually" filling this table as it composed of the rifsh_table generated from retrieve_lengths and another table that I'm manually filling from literature. Do you think that it would be better to add another step into pipeline to merge these two tables and ingest them into the airtable directly? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

@efcaguab where do you think is better to keep the manually filled table (the one I'm generating from literature data), google sheet?

efcaguab commented:
Hmm I would say the Airtable is the best place for the manual data. 

> On 1/08/2021, at 06:58, langbart ***@***.***> wrote:
> 
> ﻿
> Yes. I think it’ll be best to merge. The reason is mainly reproducibility. First, if we add more species the code wouldn’t really work unless we update the one in Airtable. Second, merging it also gives us flexibility to change the approach if needed. Lastly it could be hard to know when or wether the Airtable has the updated info from retrieve lengths.
> …
> On 31/07/2021, at 03:50, langbart @.***> wrote: ﻿ @efcaguab thanks for your review, I was just wondering how to deal with the morphometric table in airtable. At the moment I'm "manually" filling this table as it composed of the rifsh_table generated from retrieve_lengths and another table that I'm manually filling from literature. Do you think that it would be better to add another step into pipeline to merge these two tables and ingest them into the airtable directly? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.
> 
> @efcaguab where do you think is better to keep the manually filled table (the one I'm generating from literature data), google sheet?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

efcaguab commented:
Thanks for updating the code so that it combines the automated and the manual part. I did a few more changes today. 

The main change is to use a function to calculate the weight rather than merging them. This makes it possible to take advantage of the length length table to use as much data as possible and hopefully improve the estimate. It also has the added benefit that we don't need custom code for taxa like DRZ and we are less likely to miss other species where this also happens. 

I also moved as much custom taxa specific logic from being hard-coded. So added a column called length_type to the airtable for species that need something else than fork or total length. That way if we add more species we could do so easily without having to edit or write custom code. 

I also made some changes to use as much info as possible from the weight-length table. Some rows had both parameters for fork and total length but we were not using them. 

Because of these changes above the median is calculated from the weight, not the parameters. I think that's OK but let me know if you think that might be problematic. The problem is that the code is super slow. If you have any ideas of how to speed up that would be cool.

Other change I made in the manual morphometrics table in Airtable is to replace LOX for SLV. 

While doing this today I spotted a few potential problems:

- The species column in the data has quite a few problems. There are codes that do not exist in the catch_types table in Airtable. What this means is that we might need to do this step after the validation. Which is annoying because we might need to add a second round of validation for the weight and related information :/
- We also might need to fix the catch_types table in Airtable. Some of them do not have a code (but I imagine they have been recorded...) so we have no way to estimate their weight as it is now. 
- Similarly we don't have a interagency code for no catch and so all these are becoming NA which is not correct. 
- Thinking about these last two problems, if we move the weight calculation to after the validation, maybe we could do the species_code translation as part of the validation so we don't need to worry about it when doing the weight calculations
- I spotted some values in the manual morphometric table that look off. Even some negative numbers. This makes me thing that maybe they are not in the correct units. We need to make sure that all of those are in the same units as in fishbase (so cm for length and grams for weight). If not we might need to manually convert them or remove them if converting them is not feasible. 

Puff. This was quite a lot! Let me know if you have any questions! I didn't anticipate this part to be so complex! Thanks for your hard work and your patience!

langbart commented:
Wow @efcaguab, that's a lot of stuff! Thanks for these revisions and improvements. I start from the end, you are right, the wl parameters related to Hirundichthys oxycephalus in the manual table are obtained by measuring the FL in mm, I'll see if it is feasible to do the conversion. As for the issue of codes I'm not sure I understand it correctly, but it seems that there are codes that do not exist neither in airtable nor in any other table ("seaweed" "205" "418" "207" "208" "200 "" 216 "" 204 "" 206 "" 201 "" 210 "" 58 "" 148 "" 39_1 "). However, if you think it is necessary we can schedule a catch-up.

efcaguab commented:
> As for the issue of codes I'm not sure I understand it correctly, but it seems that there are codes that do not exist neither in airtable nor in any other table ("seaweed" "205" "418" "207" "208" "200 "" 216 "" 204 "" 206 "" 201 "" 210 "" 58 "" 148 "" 39_1 "). 

Yes that's what I meant. Either There are data entry errors/typos, which means we need to validate them prior OR the catch_type table might have missing types. Maybe both. We also need the codes for SLV, SWX, and CUX. @altilley or @jlopestl should be able to help here. Not sure if they get notifications from GitHub, but if not we might need to reach out to them in Teams to clarify. 

>  The wl parameters related to Hirundichthys oxycephalus in the manual table are obtained by measuring the FL in mm, I'll see if it is feasible to do the conversion.

Thanks! That would be awesome! Also, are we confident that all others are in cm/gr?

We can have a chat about it just before or after our weekly meeting tomorrow. Or later today (early tomorrow for you) if we happen to be online at the same time. :)

langbart commented:
@efcaguab I have converted the mm-derived parameters according to https://www.fishbase.de/manual/English/FishbaseThe_LENGTH_WEIGHT_table.htm. Others are all in cm/gr so we are happy. I was out this morning but yes we can have a chat about it tomorrow. For what concerns the codes, according also to the older metadata it seems they are data entry errors, and luckily they are few.

langbart commented:
@efcaguab I have just updated some catch_number codes in airtable, namely Lobster: 56  Seaweed: 58 and Sea cucumber: 57. In recent landings we are probably ok except that there are some surveys where species is labelled as "seaweed" instead of the associate code "58". In legacy landings, I have spotted one error in the recoding function (sorry for that!) for which I made a pull request. Nevertheless, there are still "205"  "418"  "207"  "208"  "200"  "216"  "204"  "206"  "201"  "210" which are very likely to be typo errors. 

Don't know if it is worths adding validation for species codes as for the moment there is only the "seaweed" issue. For what concerns the unknown codes in legacy landings I think they should be dropped unless Alex has the solution.

efcaguab commented:
Awesome! Thanks for having a look! Sounds good re. keeping it without validation and happy to keep those unknown ones as NA for now. Thanks for checking the numbers in the metadata tables (both the catch_types and the morphometrics) as well as fixing the problems with the legacy landings! 

By all means have a look at the other species name column to see if that will give us some indication but no need to delve too deep into it. If it's a substantial amount of work its also OK if we create an issue and deal with that later. 

Also let me know if you find any problems with the code I wrote. Once we iron these last bits I think (finally!) we'll be pretty close to merging this branch!

langbart commented:
Hey @efcaguab I had a look at other_species_name column and, especially in legacy landings data there is some stuff that could be processed. For example there are some entries that looks like: "46) Minor Bongko", "103) koku mutin", "114) Lanjara". So we could extract the number and replace it in the species column. Don't know yet if it is a "safe" operation. For what concerns the code it is great (as your usual), I will give a further check tomorrow
